crates/uv-python/src/discovery.rs

```diff
@@ -1150,6 +1155,24 @@ impl PythonRequest {
     pub(crate) fn is_explicit_system(&self) -> bool {
         matches!(self, Self::File(_) | Self::Directory(_))
     }
+
+    /// Serialize the request to a canonical representation.
+    ///
+    /// [`Self::parse`] should always return the same request when given the output of this method.
+    pub fn to_canonical_string(&self) -> String {
+        match self {
+            Self::Any => "any".to_string(),
+            Self::Version(version) => version.to_string(),
+            Self::Directory(path) => path.display().to_string(),
+            Self::File(path) => path.display().to_string(),
+            Self::ExecutableName(name) => name.clone(),
+            Self::Implementation(implementation) => implementation.to_string(),
+            Self::ImplementationVersion(implementation, version) => {
+                format!("{implementation}@{version}")
+            }
+            Self::Key(request) => request.to_string(),
```

Comment:
I kind of feel like this should be implementing serialize and deserialize... But I don't know what the "backend" would be, like it's not JSON etc.

Reply:
I'm a little naive about that but would be basically fine with `parse` and `to_canonical_string` moving to `Serialize` and `Deserialize` implementations. The only caveat is that `parse` is infallible.

I also considered moving to the `from_str` and `to_string` implementations and using some other helper like "user_display" for the cases where we want to show a pretty name to users?